### DISCUSSION TITLE
Add a way to use images in content collections

### DISCUSSION THREAD

**Discussion by godalming123** [withastro/roadmap#804] [Proposal]
### Body

# Summary
Each collection entry (whether it is an item in the list that `getCollection` returns, or the return value of `getEntry`) should have:
- a `src` property so that you can link to the source of a piece of content
- a `data.name` property which is generated for files that do not specify their name in their front matter by removing the extension, and just showing the file name
- a `data.type` property which is `img`, `video`, `audio`, `content` or `data` to be able to create the appropriate HTML element
- a `data.processedSrc` property which would be equivalent to the object you would get from importing an image (EG: `import myBackground from "../background.png"`)

# Background & Motivation

- I want to create a gallery where the images are organised into a folder structure to make them easier to edit
- I could not find any examples in the docs of storing images/videos/audio with collections (https://docs.astro.build/en/guides/content-collections/)
- When looking for related issues I found https://github.com/withastro/roadmap/discussions/415#discussioncomment-4461043 which says:
  > Files other than your .md/.mdx content are not permitted inside src/content/
  
  Although I think that this is out of date since the docs now say
  > This includes a type (introduced in Astro v2.5.0) specifying whether the collection contains content authoring formats like Markdown (type: 'content') or data formats like JSON or YAML (type: 'data'). It also includes a schema that defines the shape of your frontmatter or entry data.
- Globbing could be used to fetch the images, however glob returns null for the value of the object field if the file is an image
- Import cannot be used with the file path found from globbing because import needs to have the file extension specefied
- You could have a content collection of markdown files that point to images, and then get optimized to be used with the `<Picuture />` component by defining a content collection schema, but this is harder to use as you have to create 2 files per image

# Goals

- I want to be able to achieve a site with mainly images and videos which could be used (for example) for a photographer

# Example

```astro
---
import { Image } from 'astro:components';
import { getCollection } from 'astro:content';
const media = await getCollection('media');
---
<ul>
  {media.map(mediaItem => (
    { if media.data.type = "img" }
       <li><Image alt={media.data.alt} src={media.data.processedSrc} /></li>
    { endif }
    { if media.data.type = "video" }
       <!-- I do not know how the html video tag works -->
       <li><video alt={media.data.name} src={media.data.src} /></li>
    { endif }
  ))}
</ul>
```

**Comment by levino** [comment]
I think this is already [possible](https://docs.astro.build/en/guides/images/#images-in-content-collections). I am using it, it works nicely.

**Comment by godalming123** [comment]
@levino the source that you have linked shows how to use images with markdown frontmatter, I want to use images without markdown so that it is easier to add/remove images without the need to create a markdown file for each image.

**Reply by levino** [reply]
You could have one markdown file with an array of images in the frontmatter:
```markdown
---
images:
  - source: './dog.jpeg'
    alt: Dog in snow
  - source: './cat.jpeg'
    alt: A cat eating a mouse
---
```
Would that not work?

**Reply by levino** [reply]
Thinking about that, should that not also be possible with data collections? I mean the "magic" is to parse the source string as an image:
```typescript
import { defineCollection, z } from "astro:content";

const blogCollection = defineCollection({
  schema: ({ image }) => z.object({ // <-- using the schemaContext.image prop
    title: z.string(),
    cover: image(), // <-- loading and optimizing the image for usage by astro components
    coverAlt: z.string(),
  }),
});

export const collections = {
  blog: blogCollection,
};
```

**Reply by godalming123** [reply]
Thanks for this solution, it's still not optimal (and I think that this proposal is a much better solution, so I will keep this discussion open) but it's good enough, and will be what I use in this project.

**Comment by godalming123** [comment]
This might be possible with [`import.meta.glob()` (read the info text in blue)](https://docs.astro.build/en/reference/api-reference/#astroglob), but unfortunately this does not work with variables so you cannot have subdirectories easily.
**Edit:** This is not possible, and the reason is now explained in the discussion description
